microsoft/vscode
# Prioritize PATH executables over local directories in terminals

Right now, when validating an executable, the check to ensure the path is a file and not a directory comes before the check for executables in PATH. This means that a sub directory of the current process's PWD will take priority over an executable in PATH, and so an error will be given saying that the specified path "is not a file or a symlink", even if there is a valid executable in PATH which was intended to be executed.

This fix reorders the checks to ensure that the executable in PATH takes priority over a sub directory of the local PWD.

Fixes #106661 (or rather the root issue behind it, which was misdiagnosed in the thread. The error was not happening because the OS couldn't find `dotnet` in PATH, but because the `dotnet` folder in the home directory (which is usually set as the PWD when starting VSCode from an application menu) took priority over the `dotnet` executable that would be found in PATH.

To test that this issue is fixed:
1. Have an executable available from $PATH, such as `python`
2. Install an extension that has a task which runs an executable ([Here is a minimal sample extension which runs `python`](https://github.com/microsoft/vscode/files/9387119/Archive.zip))
3. Start VSCode by running `code .` in a directory which contains a sub directory with the same name as the executable (in this case, `python`)
4. Run the task which runs the executable (called `test-task` in the sample extension I have uploaded)
5. Validate that the python interactive cli starts, and there is no `The terminal process failed to launch: Path to shell executable "python" is not a file or a symlink.` error.

Also properly fixes #79346, which is the same issue.

Also fixes #83772

The steps to test this issue are the same as specified in the original issue (which can be replaced with any executable):
> 1. Have dotnet in $PATH
> 2. Open a folder
> 3. Create a dotnet folder inside the root
> 4. Set the terminal shell as dotnet

And then of course now verify that the terminal opens and there is no: `The terminal process failed to launch: Path to shell executable "dotnet" is not a file or a symlink.` error popup.
